src/classes/RD_RecurringDonations.cls

```diff
@@ -795,77 +767,391 @@ public class RD_RecurringDonations {
+            List<Opportunity> opps = [SELECT Id
+                    , Amount
+                    , Name
+                    , CampaignId
+                    , AccountID
+                    , CloseDate
+                    , StageName
+                    , isWon
+                    , Recurring_Donation_Installment_Number__c
+                    , npe03__Recurring_Donation__c
+            FROM Opportunity
+            WHERE npe03__Recurring_Donation__c IN :fixedLengthRDsToRecalculate
+            ORDER BY CloseDate ASC];
+            handleFixedLengthScheduleRecalculation(fixedLengthRDsToRecalculate, opps, dmlWrapper);
+        }
+    }
+
+    private static void handleOpenEndedScheduleRecalculation(List<npe03__Recurring_Donation__c> recurringDonations
+            , List<Opportunity> openOpportunities
+            , TDTM_Runnable.DmlWrapper dmlWrapper) {
+
+        npe03__Recurring_Donations_Settings__c rds = UTIL_CustomSettingsFacade.getRecurringDonationsSettings();
+
+        //build a map of recurring donation ID and opps
+        Map<Id, List<Opportunity>> RDOppMap = new Map<Id, List<Opportunity>>();
+        for (Opportunity o : openOpportunities){
+            if (RDOppMap.containsKey(o.npe03__Recurring_Donation__c)) {
+                List<Opportunity> opps = RDOppMap.get(o.npe03__Recurring_Donation__c);
+                opps.add(o);
+                RDOppMap.put(o.npe03__Recurring_Donation__c, opps);
             }
-            else{
-                set<id> rdset = new set<id>();
-                for (npe03__Recurring_Donation__c r : reinsertopps)
-                    rdset.add(r.id);
-                insertOppsOnRecurringDonationInsertFuture(rdset);   
+            else {
+                RDOppMap.put(o.npe03__Recurring_Donation__c, new List<Opportunity>{o});
+            }
+        }
+        // add to the map any RDs that came in with no Opportunities
+        for (npe03__Recurring_Donation__c rd : recurringDonations) {
+            if (!RDOppMap.containsKey(rd.Id)) {
+                RDOppMap.put(rd.Id, new List<Opportunity>());
+            }
+        }
+
+        List<Opportunity> updateOppsList = new List<Opportunity>();
+        List<Opportunity> closeOppsList = new List<Opportunity>();
+        List<Opportunity> insertOppsList = new List<Opportunity>();
+
+        integer forecast_months = (integer)rds.npe03__Opportunity_Forecast_Months__c;
+        Date dtForecast =  system.today().toStartOfMonth().addMonths(forecast_months);
+
+        // for each RD
+        for (npe03__Recurring_Donation__c r : recurringDonations) {
+
+            Date nextDate = getStartDate(r);
+
+            integer rdcounter = 1;
+            if (r.npe03__Total_Paid_Installments__c > 0) {
+                rdcounter = (integer)r.npe03__Total_Paid_Installments__c + 1;
+            }
+
+            // loop the existing opportunities
+            //if(RDOppMap.get(r.Id) != null && !RDOppMap.get(r.Id).isEmpty()) {
```

Comment:
i'm not a big fan of keeping commented out code that is no longer relevant.  since you put in empty lists for RD's with no open opps, i'm guessing this comment isn't needed?